### PR TITLE
Typecheck hyperparameter p3

### PR DESCRIPTION
**Pull request by smohsinali**
Remaing mypy errors in hyperparameters.py file: (mixin and value_dict)

hyperparameters.py:208: error: "is_legal" undefined in superclass
hyperparameters.py:211: error: "UniformMixin" has no attribute "upper"
hyperparameters.py:211: error: "UniformMixin" has no attribute "lower"
hyperparameters.py: note: In member "check_default" of class "UniformMixin":
hyperparameters.py:222: error: "check_default" undefined in superclass
hyperparameters.py: note: In member "__init__" of class "OrdinalHyperparameter":
hyperparameters.py:761: error: Need type annotation for variable
hyperparameters.py:764: error: Cannot determine type of 'value_dict'
hyperparameters.py: note: In member "get_order" of class "OrdinalHyperparameter":
hyperparameters.py:829: error: Cannot determine type of 'value_dict'
hyperparameters.py: note: In member "get_value" of class "OrdinalHyperparameter":
hyperparameters.py:836: error: Cannot determine type of 'value_dict'

### CODE REVIEW

**Comment by mfeurer** [comment]
I just pushed some fixes, please fix the remaining bugs. Code looks okay, but needs a little bit of restructuring and cleaning after you fixed the bugs ;)

File: ConfigSpace/hyperparameters.py [hunk 1]
```diff
@@ -30,36 +30,123 @@
...
 import io
-from functools import reduce
+import numpy as np
+
+
+def is_legal_uniformfloat(value: Union[float], upper: Union[float], lower: Union[int, float]) -> bool:
+    if not (isinstance(value, float) or isinstance(value, int)):
+        return False
+    # Strange numerical issues!
+    # todo: discuss acceptable tolerance (without big tolerance for upper the test fails)
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Tolerance is fine for now, please remove the todo and other comment.

File: ConfigSpace/hyperparameters.py [hunk 2]
```diff
@@ -30,36 +30,123 @@
...
 
 from collections import OrderedDict
-
-import numpy as np
+from typing import List, Any, Dict, Union, Tuple
 import io
-from functools import reduce
+import numpy as np
+
+
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Actually, all these methods are specific to one class, right? If yes, could you please move them to the respective classes?

File: ConfigSpace/hyperparameters.py [hunk 3]
```diff
@@ -152,91 +239,90 @@ class UnParametrizedHyperparameter(Constant):
...
         return True
 
-    def get_num_neighbors(self):
+    def get_num_neighbors(self, value: None=None) -> np.inf:
         return np.inf
 
+
 class FloatHyperparameter(NumericalHyperparameter):
-    def __init__(self, name, default):
+    # todo : type of name and default?
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Yup, they're correct. Please remove the comment.

File: ConfigSpace/hyperparameters.py [hunk 4]
```diff
@@ -152,91 +239,90 @@ class UnParametrizedHyperparameter(Constant):
...
         super(FloatHyperparameter, self).__init__(name, default)
 
-    def is_legal(self, value):
-        return isinstance(value, float) or isinstance(value, int)
-
-    def check_default(self, default):
-        return np.round(float(default), 10)
+    @abstractmethod
+    def is_legal(self, value: Union[int, float]) -> bool:
+        # return isinstance(value, float) or isinstance(value, int)
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Please remove old return statements.

File: ConfigSpace/hyperparameters.py [hunk 5]
```diff
@@ -152,91 +239,90 @@ class UnParametrizedHyperparameter(Constant):
...
+        raise NotImplemented
 
+    # todo : recheck default
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Type correct. Please remove comment.

File: ConfigSpace/hyperparameters.py [hunk 6]
```diff
@@ -152,91 +239,90 @@ class UnParametrizedHyperparameter(Constant):
...
+    def check_default(self, default: Union[int, float]) -> float:
+        # return np.round(float(default), 10)
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Please remove old return statements.

File: ConfigSpace/hyperparameters.py [hunk 7]
```diff
@@ -152,91 +239,90 @@ class UnParametrizedHyperparameter(Constant):
...
+        raise NotImplemented
 
 class IntegerHyperparameter(NumericalHyperparameter):
-    def __init__(self, name, default):
+    # todo : type of name and default?
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Correct, please remove comment.

File: ConfigSpace/hyperparameters.py [hunk 8]
```diff
@@ -152,91 +239,90 @@ class UnParametrizedHyperparameter(Constant):
...
+    def is_legal(self, value: int) -> bool:
+        raise NotImplemented
+
+    def check_int(self, parameter: int, name: str) -> int:
+        raise NotImplemented
+
+    def check_default(self, default: int) -> int:
+        raise NotImplemented
+
+# # todo: find out purpose of mixin and annotate it?
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Please remove this comments.

File: ConfigSpace/hyperparameters.py [hunk 9]
```diff
@@ -291,18 +378,23 @@ def __eq__(self, other):
         else:
             return False
 
-    def to_integer(self):
-        # TODO check if conversion makes sense at all (at least two integer
-        # values possible!)
-        return UniformIntegerHyperparameter(self.name, self.lower,
-                                            self.upper,
-                                            int(np.round(self.default)), self.q,
+    def is_legal(self, value: Union[int, float]):
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: The return type is missing here.

File: ConfigSpace/hyperparameters.py [hunk 10]
```diff
@@ -291,18 +378,23 @@ def __eq__(self, other):
...
+    def to_integer(self) -> 'UniformIntegerHyperparameter':
+        # TODO check if conversion makes sense at all (at least two integer values possible!)
+        # todo check if params should be converted to int while class initialization or inside class itself
+        return UniformIntegerHyperparameter(self.name, int(self.lower),
+                                            int(self.upper),
+                                            int(np.round(self.default)), int(self.q),
                                             self.log)
 
-    def _sample(self, rs, size=None):
+    # todo: rs? prabably numpy.random.uniform
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: No, it's `np.random.RandomState`. Please change.

File: ConfigSpace/hyperparameters.py [hunk 11]
```diff
@@ -333,16 +425,19 @@ def get_neighbors(self, value, rs, number=4, transform=False):
...
+    def __init__(self, name: str, mu: Union[int, float], sigma: Union[int, float],
+                 default: Union[None, float] = None, q: Union[int, float, None] = None, log: bool = False) -> None:
         self.mu = float(mu)
         self.sigma = float(sigma)
         self.q = float(q) if q is not None else None
         self.log = bool(log)
-        super(NormalFloatHyperparameter, self). \
-            __init__(name, self.check_default(default))
+        self.name = name
+        # super(NormalFloatHyperparameter, self). \
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Please remove this comment.

File: ConfigSpace/hyperparameters.py [hunk 12]
```diff
@@ -366,32 +461,35 @@ def __eq__(self, other):
...
 
-    def is_legal(self, value):
+    def is_legal(self, value: Union[float, int]) -> bool:
         if isinstance(value, (float, int)):
             return True
         else:
             return False
 
-    def _sample(self, rs, size=None):
+    def _sample(self, rs: np.random, size: Union[None, int] = None) -> np.ndarray:
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: rs same as above.

File: ConfigSpace/hyperparameters.py [hunk 13]
```diff
@@ -400,43 +498,42 @@ def _transform(self, vector):
...
+    def _inverse_transform(self, vector: Union[None, np.ndarray]) -> Union[float, np.ndarray]:
         if vector is None:
             return np.NaN
 
         if self.log:
             vector = np.log(vector)
         return vector
 
-    def get_neighbors(self, value, rs, number=4):
+    def get_neighbors(self, value: float, rs: np.random, number: int = 4, transform: bool = False) -> List[float]:
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: rs same as above

File: ConfigSpace/hyperparameters.py [hunk 14]
```diff
@@ -446,16 +543,17 @@ def __init__(self, name, lower, upper, default=None, q=None, log=False):
                              "hyperparameter %s is forbidden." %
                              (self.lower, name))
 
-        super(UniformIntegerHyperparameter, self). \
-            __init__(name, self.check_default(default))
+        # super(UniformIntegerHyperparameter, self). \
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Please remove unused code.

File: ConfigSpace/hyperparameters.py [hunk 15]
```diff
@@ -467,7 +565,8 @@ def __repr__(self):
         repr_str.seek(0)
         return repr_str.getvalue()
 
-    def _sample(self, rs, size=None):
+    # todo: recheck
+    def _sample(self, rs: np.random, size: Union[int, None] = None) -> np.ndarray:
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: rs same as above.

File: ConfigSpace/hyperparameters.py [hunk 16]
```diff
@@ -505,13 +608,15 @@ def has_neighbors(self):
         else:
             return False
 
-    def get_neighbors(self, value, rs, number=4, transform=False):
-        neighbors = []
+    def get_neighbors(self, value: Union[int, float], rs: np.random, number: int = 4, transform: bool = False) -> List[
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: rs same as above.

File: ConfigSpace/hyperparameters.py [hunk 17]
```diff
@@ -579,20 +688,21 @@ def __eq__(self, other):
         else:
             return False
 
-    def to_uniform(self, z=3):
+    # todo check if conversion should be done in initiation call or inside class itsel
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: At least it should be np.round instead of int. Otherwise it's probably fine here.

File: ConfigSpace/hyperparameters.py [hunk 18]
```diff
@@ -579,20 +688,21 @@ def __eq__(self, other):
...
 
-    def is_legal(self, value):
+    def is_legal(self, value: int) -> bool:
         if isinstance(value, int):
             return True
         else:
             return False
 
-    def _sample(self, rs, size=None):
+    def _sample(self, rs: np.random, size: Union[int, None] = None) -> np.ndarray:
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: rs same as above.

File: ConfigSpace/hyperparameters.py [hunk 19]
```diff
@@ -610,14 +720,16 @@ def _transform(self, vector):
...
+    def _inverse_transform(self, vector: np.ndarray) -> np.ndarray:
         return self.nfhp._inverse_transform(vector)
 
-    def has_neighbors(self):
+    def has_neighbors(self) -> bool:
         return True
 
-    def get_neighbors(self, value, rs, number=4, transform=False):
-        neighbors = []
+    # todo : find whay doesnt this function return anything
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: probably a bug. Please remove the comment.

File: ConfigSpace/hyperparameters.py [hunk 20]
```diff
@@ -659,25 +773,26 @@ def __repr__(self):
...
+    def check_default(self, default: Union[None, str, float, int]) -> Union[str, float, int]:
         if default is None:
             return self.choices[0]
         elif self.is_legal(default):
             return default
         else:
             raise ValueError("Illegal default value %s" % str(default))
 
-    def _sample(self, rs, size=None):
+    def _sample(self, rs: np.random, size: int = None) -> Union[int, np.ndarray]:
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: rs same as above.

File: ConfigSpace/hyperparameters.py [hunk 21]
```diff
@@ -659,25 +773,26 @@ def __repr__(self):
...
         return rs.randint(0, self._num_choices, size=size)
 
-    def _transform(self, vector):
-        if vector != vector:
+    # todo recheck
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: looks correct. Please remove comment.

File: ConfigSpace/hyperparameters.py [hunk 22]
```diff
@@ -686,19 +801,21 @@ def _transform(self, vector):
                              'hyperparameter %s with an integer, but provided '
                              'the following float: %f' % (self, vector))
 
-    def _inverse_transform(self, vector):
+    # todo recheck
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: looks correct. Please remove comment.

File: ConfigSpace/hyperparameters.py [hunk 23]
```diff
@@ -730,26 +847,28 @@ def get_neighbors(self, value, rs, number=np.inf, transform=False):
...
         numeric sequence according to their order/position.
         """
         super(OrdinalHyperparameter, self).__init__(name)
-        self.sequence= sequence
+        self.sequence = sequence
         self._num_elements = len(sequence)
         self.default = self.check_default(default)
-        
-        self.value_dict = OrderedDict()
+        # todo recheck type of sequence
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: looks correct, please remove comment.

File: ConfigSpace/hyperparameters.py [hunk 24]
```diff
@@ -783,8 +902,9 @@ def check_default(self, default):
             return default
         else:
             raise ValueError("Illegal default value %s" % str(default))
-            
-    def _transform(self, vector):
+
+    # todo recheck return type...is it list or normal
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: I don't get this comment.

File: ConfigSpace/hyperparameters.py [hunk 25]
```diff
@@ -794,31 +914,32 @@ def _transform(self, vector):
...
+
+    def get_order(self, value: Union[None, int, str, float]) -> int:
         """
         returns the seuence position/order of a certain value from the sequence
         """
         return self.value_dict[value]
-        
-    def get_value(self, idx):
+
+    # todo : recheck
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: correct, please remove comment.

File: ConfigSpace/hyperparameters.py [hunk 26]
```diff
@@ -829,36 +950,38 @@ def check_order(self,val1, val2):
         else:
             return False
 
-    def _sample(self, rs, size=None):
+    def _sample(self, rs: np.random, size: Union[int, None] = None) -> int:
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: same as above.

File: ConfigSpace/hyperparameters.py [hunk 27]
```diff
@@ -829,36 +950,38 @@ def check_order(self,val1, val2):
...
         returns the number of existing neighbors in the sequence
         """
-        if value == self.sequence[0] or value ==self.sequence[-1]:
+        if value == self.sequence[0] or value == self.sequence[-1]:
             return 1
         else:
             return 2
 
-    def get_neighbors(self, value, number=2, transform = False):
+    # todo: recehck...added rs as param otherrwise mismatch with baseclass signature
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Yep, correct, although it's not used here. Please remove the comment.

File: test/test_hyperparameters.py [hunk 28]
```diff
@@ -125,13 +127,22 @@ def test_uniformfloat_to_integer(self):
...
-        self.assertTrue(f1.is_legal(3))
-        self.assertFalse(f1.is_legal(-0.1))
-        self.assertFalse(f1.is_legal(10.1))
-        self.assertFalse(f1.is_legal("AAA"))
-        self.assertFalse(f1.is_legal(dict()))
+        lower = 0.1
+        upper = 10
+        f1 = UniformFloatHyperparameter("param", lower, upper, q=0.1, log=True)
+
+        # self.assertTrue(f1.is_legal(3.0))
```
Comment (mfeurer) [inline test/test_hyperparameters.py]: Why are these all commented out?

File: ConfigSpace/hyperparameters.py (hunk 2 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: both io and np are used in all classes

File: ConfigSpace/hyperparameters.py (hunk 1 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 3 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 4 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 6 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 7 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 8 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 9 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: fixed

File: ConfigSpace/hyperparameters.py (hunk 11 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 12 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: do you mean that I should not annotate rs?

File: ConfigSpace/hyperparameters.py (hunk 13 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: ?

File: ConfigSpace/hyperparameters.py (hunk 14 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 17 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: changed type of z from int to np.round

File: ConfigSpace/hyperparameters.py (hunk 19 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 21 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 22 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 24 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: outdated comment...removed

File: ConfigSpace/hyperparameters.py (hunk 25 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: ConfigSpace/hyperparameters.py (hunk 27 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: removed

File: test/test_hyperparameters.py (hunk 28 above)
Comment (smohsinali) [inline test/test_hyperparameters.py]: is_legal was changed to is_legal_uniformfloat but is now reverted back

File: ConfigSpace/hyperparameters.py [hunk 29]
```diff
@@ -239,85 +150,56 @@ class UnParametrizedHyperparameter(Constant):
...
-    # todo : type of name and default?
     def __init__(self, name: str, default: Any) -> None:
         super(NumericalHyperparameter, self).__init__(name)
         self.default = default
 
     def has_neighbors(self) -> bool:
         return True
 
-    def get_num_neighbors(self, value: None=None) -> np.inf:
+    def get_num_neighbors(self, value= None) -> np.inf:
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: Why did you remove the type declaration here?

File: ConfigSpace/hyperparameters.py [hunk 30]
```diff
@@ -239,85 +150,56 @@ class UnParametrizedHyperparameter(Constant):
...
     @abstractmethod
     def is_legal(self, value: int) -> bool:
         raise NotImplemented
 
-    def check_int(self, parameter: int, name: str) -> int:
-        raise NotImplemented
-
-    def check_default(self, default: int) -> int:
+    @abstractmethod
+    def check_default(self, default) -> int:
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: No type definition here.

File: ConfigSpace/hyperparameters.py [hunk 31]
```diff
@@ -689,18 +605,24 @@ def __eq__(self, other: Any) -> bool:
             return False
 
     # todo check if conversion should be done in initiation call or inside class itsel
-    def to_uniform(self, z: int = 3) -> 'UniformIntegerHyperparameter':
+    def to_uniform(self, z: np.round = 3) -> 'UniformIntegerHyperparameter':
```
Comment (mfeurer) [inline ConfigSpace/hyperparameters.py]: I didn't mean the type should be `np.round`, I meant you should use `np.round(int(self.mu - (z * self.sigma)))` instead of only `int(self.mu - (z * self.sigma))` to make sure that the hyperparameter is rounded correctly to the next int.

Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: fixed

File: ConfigSpace/hyperparameters.py (hunk 29 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: mypy doesnt accept None=None

File: ConfigSpace/hyperparameters.py (hunk 30 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: this is abstract method. in its definition in UniformIntegerHyperparameter type of default is 'Union[int, float]' while in NormalIntegerHyperparameter type of default is 'int' so because of this conflict i removed it form abstract method.

if you want in some different way i can do that for example making it same and then in __init__ of NormalIntegerHyperparameter doing a check to make sure that its only 'int'

File: ConfigSpace/hyperparameters.py (hunk 13 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: changed to np.random.RandomState

File: ConfigSpace/hyperparameters.py (hunk 16 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: changed to np.random.RandomState

File: ConfigSpace/hyperparameters.py (hunk 20 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: changed to np.random.RandomState

File: ConfigSpace/hyperparameters.py (hunk 26 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: changed to np.random.RandomState

File: ConfigSpace/hyperparameters.py (hunk 31 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: fixed

File: ConfigSpace/hyperparameters.py (hunk 10 above)
Comment (smohsinali) [inline ConfigSpace/hyperparameters.py]: fxed
